p2panda-rs/src/document/document.rs

```diff
@@ -0,0 +1,406 @@
+        let sorted = match self.graph.sort() {
+            Ok(operations) => Ok(operations.into_iter().rev().collect()),
+            Err(e) => Err(DocumentError::GraphSortingError(e)),
+        }?;
+
+        Ok(DocumentIter(sorted))
+    }
+
+    /// Sort the graph topologically, then reduce the linearised operations into a single
+    /// `Instance`.
+    pub fn resolve(&self) -> Result<Instance, DocumentError> {
+        let mut document_iter = self.iter()?;
+
+        let create_message = document_iter.next().unwrap();
+        let mut instance = Instance::try_from(create_message)?;
+
+        document_iter.try_for_each(|op| instance.apply_update(op))?;
+
+        Ok(instance)
+    }
+}
+
+impl Iterator for DocumentIter {
+    type Item = OperationWithMeta;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.pop()
+    }
+}
+
+/// A struct for building documents.
+#[derive(Debug)]
+pub struct DocumentBuilder {
+    /// An unsorted collection of operations which are associated with a particular document id.
+    operations: Vec<OperationWithMeta>,
+    schema_definition: String,
+}
+
+impl DocumentBuilder {
+    /// Instantiate a new DocumentBuilder with a collection of operations.
+    pub fn new(operations: Vec<OperationWithMeta>, schema_definition: String) -> Self {
+        Self {
+            operations,
+            schema_definition,
+        }
+    }
+
+    /// Get all operations for this document.
+    pub fn operations(&self) -> Vec<OperationWithMeta> {
+        self.operations.clone()
+    }
+
+    /// Get an iterator over all operations in this document.
+    pub fn operations_iter(&self) -> std::vec::IntoIter<OperationWithMeta> {
+        self.operations.clone().into_iter()
+    }
+
+    /// Build the document.
+    pub fn build(self) -> Result<Document, DocumentBuilderError> {
+        // find create message.
+        let collect_create_operation: Vec<OperationWithMeta> =
+            self.operations_iter().filter(|op| op.is_create()).collect();
+
+        // Check we have only one create operation in the document.
+        if collect_create_operation.len() > 1 {
+            return Err(DocumentBuilderError::MoreThanOneCreateOperation);
+        } else if collect_create_operation.is_empty() {
+            return Err(DocumentBuilderError::NoCreateOperation);
+        }
+
+        let create_operation = &collect_create_operation[0];
+
+        let document_schema = create_operation.schema();
+
+        // Create instantiate a schema, which validates the provided schema's CDDL definition.
+        let schema = Schema::new(&create_operation.schema(), &self.schema_definition)?;
+
+        // Validate the create message for this document against the document schema.
+        // @TODO: Exclude from wasm target until cddl issues sorted (https://github.com/p2panda/p2panda/issues/99).
+        #[cfg(not(target_arch = "wasm32"))]
+        schema.validate_operation_fields(&create_operation.fields().unwrap())?;
```

Comment:
Suggestion to remove this validation as we already know that operations conform to their schema